File: backend/app/api/v1/loans.py
import uuid
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_session
from app.core.deps import PaginationParams, get_current_user
from app.crud.loan import (
    create_loan,
    get_loan,
    list_loans,
    soft_delete_loan,
    update_loan,
)
from app.models.user import User
from app.schemas.loan import LoanCreate, LoanPublic, LoanUpdate
from app.schemas.paginated import Paginated

router = APIRouter(prefix="/loans", tags=["loans"])
person_loans_router = APIRouter(prefix="/persons/{person_id}/loans", tags=["loans"])


@router.post("/", response_model=LoanPublic, status_code=status.HTTP_201_CREATED)
async def create(
    data: LoanCreate,
    db: Annotated[AsyncSession, Depends(get_session)],
    current_user: Annotated[User, Depends(get_current_user)],
):
    return await create_loan(db, current_user.id, data)


@router.get("/", response_model=Paginated[LoanPublic])
async def list_all(
    db: Annotated[AsyncSession, Depends(get_session)],
    current_user: Annotated[User, Depends(get_current_user)],
    pagination: Annotated[PaginationParams, Depends(PaginationParams)],
    status_filter: str | None = None,
    direction: str | None = None,
):
    items, total = await list_loans(
        db,
        current_user.id,
        skip=pagination.skip,
        limit=pagination.limit,
        status=status_filter,
        direction=direction,
    )
    return Paginated(
        items=items, total=total, skip=pagination.skip, limit=pagination.limit
    )


@router.get("/{loan_id}", response_model=LoanPublic)
async def get_one(
    loan_id: uuid.UUID,
    db: Annotated[AsyncSession, Depends(get_session)],
    current_user: Annotated[User, Depends(get_current_user)],
):
    loan = await get_loan(db, loan_id, current_user.id)
    if not loan:
        raise HTTPException(status_code=404, detail="Loan not found")
    return loan


@router.patch("/{loan_id}", response_model=LoanPublic)
async def patch(
    loan_id: uuid.UUID,
    data: LoanUpdate,
    db: Annotated[AsyncSession, Depends(get_session)],
    current_user: Annotated[User, Depends(get_current_user)],
):
    loan = await update_loan(db, loan_id, current_user.id, data)
    if not loan:
        raise HTTPException(status_code=404, detail="Loan not found")
    return loan


@router.delete("/{loan_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete(
    loan_id: uuid.UUID,
    db: Annotated[AsyncSession, Depends(get_session)],
    current_user: Annotated[User, Depends(get_current_user)],
):
    deleted = await soft_delete_loan(db, loan_id, current_user.id)
    if not deleted:
        raise HTTPException(status_code=404, detail="Loan not found")


# ── Person-scoped loans ──────────────────────────────────────────────────────


@person_loans_router.get("/", response_model=Paginated[LoanPublic])
async def list_person_loans(
    person_id: uuid.UUID,
    db: Annotated[AsyncSession, Depends(get_session)],
    current_user: Annotated[User, Depends(get_current_user)],
    pagination: Annotated[PaginationParams, Depends(PaginationParams)],
):
    items, total = await list_loans(
        db,
        current_user.id,
        skip=pagination.skip,
        limit=pagination.limit,
        person_id=person_id,
    )
    return Paginated(
        items=items, total=total, skip=pagination.skip, limit=pagination.limit
    )

File: frontend/src/api/loans.ts
import { api } from "./client";
import type { Paginated } from "./types";

export interface LoanPublic {
  id: string;
  owner_id: string;
  person_id: string;
  direction: string;
  loan_type: string;
  description: string;
  amount: number | null;
  currency: string | null;
  item_name: string | null;
  loaned_on: string | null;
  due_on: string | null;
  returned_on: string | null;
  status: string;
  notes: string | null;
  created_at: string;
  updated_at: string;
}

export interface LoanCreatePayload {
  person_id: string;
  direction: string;
  loan_type: string;
  description: string;
  amount?: number | null;
  currency?: string | null;
  item_name?: string | null;
  loaned_on?: string | null;
  due_on?: string | null;
  notes?: string | null;
}

export interface LoanUpdatePayload {
  description?: string;
  amount?: number | null;
  currency?: string | null;
  item_name?: string | null;
  loaned_on?: string | null;
  due_on?: string | null;
  returned_on?: string | null;
  status?: string;
  notes?: string | null;
}

export function listLoans(params: {
  skip?: number;
  limit?: number;
  direction?: string;
  status_filter?: string;
} = {}): Promise<Paginated<LoanPublic>> {
  const q = new URLSearchParams();
  if (params.skip !== undefined) q.set("skip", String(params.skip));
  if (params.limit !== undefined) q.set("limit", String(params.limit));
  if (params.direction) q.set("direction", params.direction);
  if (params.status_filter) q.set("status_filter", params.status_filter);
  const qs = q.toString();
  return api.get<Paginated<LoanPublic>>(`/loans${qs ? `?${qs}` : ""}`);
}

export function getLoan(id: string): Promise<LoanPublic> {
  return api.get<LoanPublic>(`/loans/${id}`);
}

export function createLoan(data: LoanCreatePayload): Promise<LoanPublic> {
  return api.post<LoanPublic>("/loans", data);
}

export function updateLoan(id: string, data: LoanUpdatePayload): Promise<LoanPublic> {
  return api.patch<LoanPublic>(`/loans/${id}`, data);
}

export function deleteLoan(id: string): Promise<void> {
  return api.delete(`/loans/${id}`);
}
